orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/poller/join/MinionAcceptor.java

```diff
@@ -36,14 +36,32 @@ public MinionAcceptor(SaltConnector sc, List<Minion> minions, MinionFingerprintM
     }
 
     public void acceptMinions() throws CloudbreakOrchestratorFailedException {
-        List<String> unacceptedMinions = fetchUnacceptedMinionsFromMaster(minions);
+        MinionKeysOnMasterResponse minionKeysOnMaster = fetchMinionsFromMaster(minions);
+        List<String> unacceptedMinions = minionKeysOnMaster.getUnacceptedMinions();
+        List<String> deniedMinions = minionKeysOnMaster.getDeniedMinions();
+        removeMinionIdsInBothDeniedAndUnacceptedState(deniedMinions, unacceptedMinions);
         if (!unacceptedMinions.isEmpty()) {
             proceedWithAcceptingMinions(unacceptedMinions);
         } else {
             LOGGER.info("No unaccepted minions found on master");
         }
     }
 
+    private void removeMinionIdsInBothDeniedAndUnacceptedState(List<String> deniedMinions, List<String> unacceptedMinions)
+            throws CloudbreakOrchestratorFailedException {
+        LOGGER.info("Unaccepted minions: {}", unacceptedMinions);
+        LOGGER.info("Denied minions: {}", deniedMinions);
+        List<String> minionIdsInBothDeniedAndUnacceptedState = deniedMinions.stream().distinct()
+                .filter(unacceptedMinions::contains).collect(Collectors.toList());
+        if (!minionIdsInBothDeniedAndUnacceptedState.isEmpty()) {
+            LOGGER.info("There are minions in denied and unaccepted state at the same time, lets remove them: " +
+                    minionIdsInBothDeniedAndUnacceptedState);
+            sc.wheel("key.delete", minionIdsInBothDeniedAndUnacceptedState, Object.class);
```

Comment:
Shouldn't we introduce a method called Wheel.deleteKey?  Similar like as we do in  SaltStates.* Ther reason for that is delete.keys is used in multiple places in the code and we could have common logging  and maybe even error handling.